### PR TITLE
infoschema/executor : add DDLJobs sys table

### PR DESCRIPTION
**Pull request by reafans**
### What problem does this PR solve? 
Currently, we can view the history of DDL jobs and DDL jobs by 'admin show DDL jobs', but it is not convenient to use queries such as group by, so we need to add a system table of DDL_JOBS.

### What is changed and how it works?
```
mysql> select * from information_schema.DDL_JOBS limit 10;
+--------+---------------+------------+---------------+--------------+-----------+----------+-----------+-----------------------------------+-----------------------------------+--------+--------------------------------------------------------------------------------------------------------------------------------------------------------+
| JOB_ID | DB_NAME       | TABLE_NAME | JOB_TYPE      | SCHEMA_STATE | SCHEMA_ID | TABLE_ID | ROW_COUNT | START_TIME                        | END_TIME                          | STATE  | QUERY                                                                                                                                                  |
+--------+---------------+------------+---------------+--------------+-----------+----------+-----------+-----------------------------------+-----------------------------------+--------+--------------------------------------------------------------------------------------------------------------------------------------------------------+
|   1718 | test_ddl_jobs |            | create schema | public       |      1717 |        0 |         0 | 2020-02-18 20:35:40.43 +0800 CST  | 2020-02-18 20:35:40.434 +0800 CST | synced | create database if not exists test_ddl_jobs                                                                                                            |
|   1716 | test_ddl_jobs |            | drop schema   | none         |      1714 |        0 |         0 | 2020-02-18 19:50:29.07 +0800 CST  | 2020-02-18 19:50:29.084 +0800 CST | synced | drop database test_ddl_jobs                                                                                                                            |
|   1715 | test_ddl_jobs |            | create schema | public       |      1714 |        0 |         0 | 2020-02-18 19:45:54.182 +0800 CST | 2020-02-18 19:45:54.185 +0800 CST | synced | create database if not exists test_ddl_jobs                                                                                                            |
|   1713 | test          | level0     | create table  | public       |         1 |     1712 |         0 | 2020-02-06 20:32:49.557 +0800 CST | 2020-02-06 20:32:49.563 +0800 CST | synced | CREATE TABLE `level0` (`id` int unsigned AUTO_INCREMENT,`value` varchar(255),`level1_id` int unsigned , PRIMARY KEY (`id`))                            |
|   1711 | test          | level2_1   | create table  | public       |         1 |     1710 |         0 | 2020-02-06 20:32:49.546 +0800 CST | 2020-02-06 20:32:49.552 +0800 CST | synced | CREATE TABLE `level2_1` (`id` int unsigned AUTO_INCREMENT,`level3_id` int unsigned , PRIMARY KEY (`id`))                                               |
|   1709 | test          | level1     | create table  | public       |         1 |     1708 |         0 | 2020-02-06 20:32:49.536 +0800 CST | 2020-02-06 20:32:49.541 +0800 CST | synced | CREATE TABLE `level1` (`id` int unsigned AUTO_INCREMENT,`value` varchar(255),`level2_id` int unsigned,`level2_1_id` int unsigned , PRIMARY KEY (`id`)) |
|   1707 | test          | level2     | create table  | public       |         1 |     1706 |         0 | 2020-02-06 20:32:49.527 +0800 CST | 2020-02-06 20:32:49.532 +0800 CST | synced | CREATE TABLE `level2` (`id` int unsigned AUTO_INCREMENT,`level3_id` int unsigned , PRIMARY KEY (`id`))                                                 |
|   1705 | test          | level3     | create table  | public       |         1 |     1704 |         0 | 2020-02-06 20:32:49.518 +0800 CST | 2020-02-06 20:32:49.523 +0800 CST | synced | CREATE TABLE `level3` (`id` int unsigned AUTO_INCREMENT,`name` varchar(255) , PRIMARY KEY (`id`))                                                      |
|   1703 | test          | level0     | drop table    | none         |         1 |     1554 |         0 | 2020-02-06 20:32:49.5 +0800 CST   | 2020-02-06 20:32:49.513 +0800 CST | synced | DROP TABLE `level0`                                                                                                                                    |
|   1702 | test          | level1     | drop table    | none         |         1 |     1697 |         0 | 2020-02-06 20:32:49.479 +0800 CST | 2020-02-06 20:32:49.493 +0800 CST | synced | DROP TABLE `level1`                                                                                                                                    |
+--------+---------------+------------+---------------+--------------+-----------+----------+-----------+-----------------------------------+-----------------------------------+--------+--------------------------------------------------------------------------------------------------------------------------------------------------------+
10 rows in set (0.00 sec)
```

Group by: 
```
mysql> select db_name,count(*) from DDL_JOBS group by db_name;
+---------------+----------+
| db_name       | count(*) |
+---------------+----------+
| test_ddl_jobs |        3 |
| test          |     1119 |
| mysql         |       23 |
+---------------+----------+
3 rows in set (0.07 sec)
```
### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Has exported variable/fields change

### CODE REVIEW

File: infoschema/tables.go [hunk 1]
```diff
@@ -839,6 +842,20 @@ var tableClusterSystemInfoCols = []columnInfo{
...
+	{"DB_NAME", mysql.TypeVarchar, 64, 0, nil, nil},
+	{"TABLE_NAME", mysql.TypeVarchar, 64, 0, nil, nil},
+	{"JOB_TYPE", mysql.TypeVarchar, 64, 0, nil, nil},
+	{"SCHEMA_STATE", mysql.TypeVarchar, 64, 0, nil, nil},
+	{"SCHEMA_ID", mysql.TypeLonglong, 4, 0, nil, nil},
+	{"TABLE_ID", mysql.TypeLonglong, 4, 0, nil, nil},
+	{"ROW_COUNT", mysql.TypeLonglong, 4, 0, nil, nil},
+	{"START_TIME", mysql.TypeVarchar, 64, 0, nil, nil},
+	{"END_TIME", mysql.TypeVarchar, 64, 0, nil, nil},
+	{"STATE", mysql.TypeVarchar, 64, 0, nil, nil},
```
Comment (zimulala) [inline infoschema/tables.go]: Could we add the column of `Query`?

Comment (reafans) [reply infoschema/tables.go]: done

File: executor/infoschema_reader.go [hunk 2]
```diff
@@ -15,6 +15,10 @@ package executor
 
 import (
 	"context"
+	"github.com/cznic/mathutil"
+	"github.com/pingcap/tidb/meta"
+	"github.com/pingcap/tidb/util/admin"
+	"github.com/pingcap/tidb/util/chunk"
```
Comment (crazycs520) [inline executor/infoschema_reader.go]: format code.
move #18 ~ #21 after line #14523

File: executor/infoschema_reader.go [hunk 3]
```diff
@@ -153,3 +157,107 @@ func dataForViews(ctx sessionctx.Context, schemas []*model.DBInfo) ([][]types.Da
...
+	}
+	e.runningJobs = append(e.runningJobs, jobs...)
+	e.cursor = 0
+	return nil
+}
+
+// Next implements the Executor Next interface.
+func (e *DDLJobsReaderExec) Next(ctx context.Context, req *chunk.Chunk) error {
+	req.GrowAndReset(e.maxChunkSize)
+	if !e.historyJobIter.Valid() {
```
Comment (crazycs520) [inline executor/infoschema_reader.go]: Why add this check? It looks redundant, any test for here?

Comment (reafans) [reply executor/infoschema_reader.go]: I checked and it is indeed redundant, removed.

Comment (reafans) [reply executor/infoschema_reader.go]: I check and it indeed is redundant, removed

File: executor/infoschema_reader.go [hunk 4]
```diff
@@ -153,3 +157,104 @@ func dataForViews(ctx sessionctx.Context, schemas []*model.DBInfo) ([][]types.Da
...
+	jobs, err := admin.GetDDLJobs(txn)
+	if err != nil {
+		return err
+	}
+	m := meta.NewMeta(txn)
+	e.historyJobIter, err = m.GetLastHistoryDDLJobsIterator()
+	if err != nil {
+		return err
+	}
+	e.runningJobs = append(e.runningJobs, jobs...)
```
Comment (lonng) [inline executor/infoschema_reader.go]: Why not assign `jobs` to `e.runningJobs` directly?

File: executor/infoschema_reader.go [hunk 5]
```diff
@@ -153,3 +157,104 @@ func dataForViews(ctx sessionctx.Context, schemas []*model.DBInfo) ([][]types.Da
 	}
 	return rows, nil
 }
+
+// DDLJobsReaderExec executes DDLJobs information retrieving
+type DDLJobsReaderExec struct {
```
Comment (djshow832) [inline executor/infoschema_reader.go]: Too much code is duplicate with `ShowDDLJobQueriesExec`, I suggest refactoring it.

File: executor/infoschema_reader_test.go [hunk 6]
```diff
@@ -83,3 +87,30 @@ func (s *testInfoschemaTableSuite) TestViews(c *C) {
 	tk.MustQuery("SELECT * FROM information_schema.views WHERE table_schema='test' AND table_name='v1'").Check(testkit.Rows("def test v1 SELECT 1 CASCADED NO root@localhost DEFINER utf8mb4 utf8mb4_bin"))
 	tk.MustQuery("SELECT table_catalog, table_schema, table_name, table_type, engine, version, row_format, table_rows, avg_row_length, data_length, max_data_length, index_length, data_free, auto_increment, update_time, check_time, table_collation, checksum, create_options, table_comment FROM information_schema.tables WHERE table_schema='test' AND table_name='v1'").Check(testkit.Rows("def test v1 VIEW <nil> <nil> <nil> <nil> <nil> <nil> <nil> <nil> <nil> <nil> <nil> <nil> <nil> <nil> <nil> VIEW"))
 }
+
+func (s *testInfoschemaTableSuite) TestDDLJobs(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("create database if not exists test_ddl_jobs")
+	re := tk.MustQuery("select * from information_schema.DDL_JOBS")
```
Comment (lonng) [inline executor/infoschema_reader_test.go]: It's better to use 
```
tk.MustQuery("select db_name, job_type from information_schema.DDL_JOBS").Check(testkit.Rows(
    "test_ddl_jobs schema"
))
```

File: executor/infoschema_reader_test.go [hunk 7]
```diff
@@ -83,3 +87,30 @@ func (s *testInfoschemaTableSuite) TestViews(c *C) {
...
+	row := re.Rows()[0]
+	c.Assert(row[1], Equals, "test_ddl_jobs")
+	c.Assert(row[3], Equals, "create schema")
+	jobID, err := strconv.Atoi(row[0].(string))
+	c.Assert(err, IsNil)
+
+	tk.MustExec("use test_ddl_jobs")
+	tk.MustExec("create table t (a int);")
+	re = tk.MustQuery("select * from information_schema.DDL_JOBS where job_type='create table'")
```
Comment (lonng) [inline executor/infoschema_reader_test.go]: ditto

File: executor/infoschema_reader_test.go (hunk 6 above)
Comment (reafans) [reply executor/infoschema_reader_test.go]: done.

File: executor/infoschema_reader_test.go (hunk 7 above)
Comment (reafans) [reply executor/infoschema_reader_test.go]: done.

File: executor/infoschema_reader.go (hunk 4 above)
Comment (reafans) [reply executor/infoschema_reader.go]: done.

File: executor/infoschema_reader.go (hunk 5 above)
Comment (reafans) [reply executor/infoschema_reader.go]: Add a struct `DDLJobExecInitializer` to reuse the DDLJobs initial logic in `Open()`, and other operation has some different code should bound to their executor which would not be refactord.

File: executor/infoschema_reader.go [hunk 8]
```diff
@@ -15,14 +15,16 @@ package executor
...
+	"github.com/cznic/mathutil"
 	"github.com/pingcap/parser/model"
 	"github.com/pingcap/parser/mysql"
 	"github.com/pingcap/tidb/infoschema"
 	"github.com/pingcap/tidb/privilege"
 	"github.com/pingcap/tidb/sessionctx"
 	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+	"sort"
+	"strings"
```
Comment (djshow832) [inline executor/infoschema_reader.go]: Format the imports.

File: meta/meta.go [hunk 9]
```diff
@@ -779,8 +779,8 @@ func (m *Meta) GetLastHistoryDDLJobsIterator() (*LastJobIterator, error) {
 	}, nil
 }
 
-// GetLastJobs gets last several jobs.
-func (i *LastJobIterator) GetLastJobs(num int, jobs []*model.Job) ([]*model.Job, error) {
+// GetNextJobs gets next several jobs.
+func (i *LastJobIterator) GetNextJobs(num int, jobs []*model.Job) ([]*model.Job, error) {
```
Comment (zimulala) [inline meta/meta.go]: Why change this function name? I think the last job is better.

Comment (reafans) [reply meta/meta.go]: Em..., last mostly means the end, ok, it also can stand the last job in the LastJobIterator.
Changed back.

File: executor/infoschema_reader.go [hunk 10]
```diff
@@ -241,6 +244,106 @@ func dataForViews(ctx sessionctx.Context, schemas []*model.DBInfo) ([][]types.Da
 	return rows, nil
 }
 
+// DDLJobsReaderExec executes DDLJobs information retrieving.
+type DDLJobsReaderExec struct {
+	baseExecutor
+	DDLJobExecInitializer
+
+	cacheJobs []*model.Job
```
Comment (crazycs520) [inline executor/infoschema_reader.go]: How about move `cacheJobs` into `DDLJobExecInitializer`?

File: executor/executor.go [hunk 11]
```diff
@@ -329,14 +329,34 @@ func (e *ShowDDLExec) Next(ctx context.Context, req *chunk.Chunk) error {
...
 
-	cursor         int
+	cacheJobs []*model.Job
+	jobNumber int
+	is        infoschema.InfoSchema
+	done      bool
+}
+
+// DDLJobExecInitializer initials the DDLJobsExec.
+type DDLJobExecInitializer struct {
```
Comment (crazycs520) [inline executor/executor.go]: No need to export `DDLJobExecInitializer`?

Comment (crazycs520) [reply executor/executor.go]: BTW, How about rename to `DDLJobRetriever`?

File: executor/infoschema_reader.go [hunk 12]
```diff
@@ -241,6 +244,106 @@ func dataForViews(ctx sessionctx.Context, schemas []*model.DBInfo) ([][]types.Da
...
+		}
+		e.cursor += len(e.cacheJobs)
+	}
+	return nil
+}
+
+func (e *DDLJobsReaderExec) appendJobToChunk(req *chunk.Chunk, job *model.Job, checker privilege.Manager) {
+	schemaName := job.SchemaName
+	tableName := ""
+	finishTS := uint64(0)
```
Comment (crazycs520) [inline executor/infoschema_reader.go]: Line #306 ~ 322 is duplicated, how about extract to function or method of DDLJobExecInitializer

File: executor/executor.go (hunk 11 above)
Comment (reafans) [reply executor/executor.go]: done

File: executor/infoschema_reader.go (hunk 12 above)
Comment (reafans) [reply executor/infoschema_reader.go]: done

File: executor/infoschema_reader_test.go [hunk 13]
```diff
@@ -14,15 +14,15 @@
 package executor_test
 
 import (
+	"github.com/pingcap/tidb/executor"
+	"github.com/pingcap/tidb/statistics/handle"
```
Comment (crazycs520) [inline executor/infoschema_reader_test.go]: format code

File: executor/infoschema_reader_test.go [hunk 14]
```diff
@@ -149,6 +149,39 @@ func (s *testInfoschemaTableSuite) TestCharacterSetCollations(c *C) {
...
+	}, nil, nil), IsTrue)
+	DDLJobsTester.MustQuery("select DB_NAME from information_schema.DDL_JOBS;").Check(
+		[][]interface{}{})
+
+	// Test the privilege of user with privilege of mysql for information_schema.ddl_jobs.
+	tk.MustExec("CREATE ROLE mysql_priv;")
+	tk.MustExec("GRANT ALL PRIVILEGES ON mysql.* TO mysql_priv;")
+	tk.MustExec("GRANT mysql_priv TO DDL_JOBS_tester;")
+	DDLJobsTester.MustExec("set role mysql_priv")
+	DDLJobsTester.MustQuery("select DISTINCT DB_NAME from information_schema.DDL_JOBS where DB_NAME = 'mysql';").Check(testkit.Rows("mysql"))
```
Comment (crazycs520) [inline executor/infoschema_reader_test.go]: I think you should test the table that the user doesn't have the privilege to see it.

Comment (reafans) [reply executor/infoschema_reader_test.go]: done

File: executor/infoschema_reader_test.go [hunk 15]
```diff
@@ -16,6 +16,7 @@ package executor_test
 import (
 	"crypto/tls"
 	"fmt"
+	"google.golang.org/grpc"
```
Comment (crazycs520) [inline executor/infoschema_reader_test.go]: format code

**Pull request merged**